spynnaker/pyNN/models/neuron/synaptic_matrix_app.py

```diff
@@ -0,0 +1,719 @@
+                undelayed_matrix_data.append((m_edge, row_data))
+                delayed_matrix_data.append((m_edge, delay_row_data))
+            else:
+                # If no app keys, write the data as normal
+                block_addr, single_addr = matrix.write_machine_matrix(
+                    spec, block_addr, single_synapses, single_addr,
+                    row_data)
+                block_addr = matrix.write_delayed_machine_matrix(
+                    spec, block_addr, delay_row_data)
+
+        # If there is an app key, add a single matrix
+        if self.__use_app_keys:
+            block_addr = self.__write_app_matrix(
+                spec, block_addr, undelayed_matrix_data)
+            block_addr = self.__write_delay_app_matrix(
+                spec, block_addr, delayed_matrix_data)
+
+        return block_addr, single_addr
+
+    def __write_app_matrix(self, spec, block_addr, matrix_data):
+        """ Write a matrix for a whole incoming application vertex as one
+
+        :param DataSpecificationGenerator spec:
+            The specification to write to
+        :param int block_addr:
+            The address in the synaptic matrix region to start writing at
+        :param list(ProjectionMachineEdge, ~numpy.ndarray) matrix_data:
+            The data for each machine edge to be combined into a single matrix
+        :return: The updated block address
+        :rtype: int
+        """
+        # If there is no routing info, don't write anything
+        if self.__app_key_info is None:
+            return block_addr
+
+        # If we have routing info but no synapses, write an invalid entry
+        if self.__max_row_info.undelayed_max_n_synapses == 0:
+            self.__index = self.__poptable.add_invalid_entry(
+                self.__app_key_info.key_and_mask)
+            return block_addr
+
+        # Write a matrix for the whole application vertex
+        block_addr = self.__poptable.write_padding(spec, block_addr)
+        self.__index = self.__poptable.add_application_entry(
+            block_addr,  self.__max_row_info.undelayed_max_words,
+            self.__app_key_info.key_and_mask, self.__app_key_info.core_mask,
+            self.__app_key_info.core_shift, self.__app_key_info.n_neurons)
+        self.__syn_mat_offset = block_addr
+
+        # Write all the row data for each machine vertex one after the other.
+        # Implicit assumption that no machine-level row_data is ever empty;
+        # this must be true in the current code, because the row length is
+        # fixed for all synaptic matrices from the same source application
+        # vertex
+        for m_edge, row_data in matrix_data:
+            size = (self.__max_row_info.undelayed_max_bytes *
+                    m_edge.pre_vertex.vertex_slice.n_atoms)
+            row_data_size = len(row_data) * BYTES_PER_WORD
+            if row_data_size != size:
+                raise Exception("Data incorrect size: {} instead of {}".format(
```

Comment:
error message at top of file is new format we seem to be going to

Reply:
I disagree; there are plenty of places we don't do that.  It is only done in some places because of a very big error message that messes up the code.  It is highly prone to error given the arguments need to match, and given that the exceptions are basically untested (even where this has been done, I would be surprised if anyone has checked).  If anything, the exception could be encapsulated in a function call that would check the argument alignment, but this exception is just too simple for that to happen.  It is basically a sanity check if nothing else, used purely for debugging.